src/lab12.py

```diff
@@ -701,21 +717,27 @@ def raster_and_draw(self):
         if self.needs_draw:
             timer = Timer()
             timer.start()
+        else:
+            assert not self.needs_chrome_raster
+            assert not self.needs_tab_raster
+            self.compositor_lock.release()
+            return
```

Comment:
I think the `assert`s here point to a todo for the future: the set of flags has at this point gotten kind of overwhelming, and the early return indicates that the code is no longer set up to match the underlying logic particularly closely. I admit that I too am a little lost as to how many flags there are and how they relate. (I can usually tell you what flags must be set from a particular location, but it's hard for me to recall which locations set a given flag or whether one flag implies another.) We should simplify if possible, or if impossible we should add some clarifying text so the reader can follow along.

Reply:
Agree we should simplify if possible.